bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthStoreHandlerImpl.java

```diff
@@ -0,0 +1,555 @@
+                    _storage.put(LAST_USED.getKey(handle), gsonDateStr);
+                    if (!allHandles.contains(handle)) {
+                        // update all handles index
+                        allHandles.add(handle);
+                        _storage.put(STORE_KEY_INDEX_OF_HANDLES, gson.toJson(allHandles));
+                    }
+                }
+            } finally {
+                storageLock.unlock();
+            }
+        }
+
+        public void removeByHandle(String handle) {
+            logger.info("Removing handle {} from storage", handle);
+            storageLock.lock();
+            try {
+                if (allHandles.remove(handle)) { // entry exists and successfully removed
+                    if (_storage == null) {
+                        // just log the error and let it be, do not throw exception
+                        logger.warn(EXCEPTION_STORAGE);
+                    } else {
+                        _storage.remove(ACCESS_TOKEN_RESPONSE.getKey(handle));
+                        _storage.remove(LAST_USED.getKey(handle));
+                        _storage.remove(SERVICE_CONFIGURATION.getKey(handle));
+                        _storage.put(STORE_KEY_INDEX_OF_HANDLES, gson.toJson(allHandles)); // update all handles
+                    }
+                }
+            } finally {
+                storageLock.unlock();
+            }
+        }
+
+        public void removeAll() {
+            // no need any locks, the other methods will take care of this
+            Set<String> allHandlesFromStore = getAllHandlesFromIndex();
+            for (String handle : allHandlesFromStore) {
+                removeByHandle(handle);
+            }
+        }
+
+        @Override
+        public void close() {
+            boolean lockGained = false;
+            try {
+                // dont want to wait too long during shutdown or update
+                lockGained = storageLock.tryLock(15, TimeUnit.SECONDS);
+
+                // if lockGained within timeout, then try to remove old entries
+                if (lockGained) {
+                    if (_storage == null) {
+                        // no resources to release/ delete, just return.
+                        return;
+                    }
+
+                    String handlesSSV = this._storage.get(STORE_KEY_INDEX_OF_HANDLES);
+                    if (handlesSSV != null) {
+                        String[] handles = handlesSSV.trim().split(" ");
+                        for (String handle : handles) {
+                            LocalDateTime lastUsed = (LocalDateTime) get(handle, LAST_USED);
+                            if (isExpired(lastUsed)) {
+                                removeByHandle(handle);
+                            }
+                        }
+                    }
+                }
+            } catch (InterruptedException e) {
+                // if lock is not acquired within the timeout or thread is interruted
```

Comment:
Thread.currentThread.interrupt()

Reply:
InterruptedExceptions should never be ignored in the code, and simply logging the exception counts in this case as "ignoring". The throwing of the InterruptedException clears the interrupted state of the Thread, so if the exception is not handled properly the fact that the thread was interrupted will be lost. Instead, InterruptedExceptions should either be rethrown - immediately or after cleaning up the method's state - or the thread should be re-interrupted by calling Thread.interrupt() even if this is supposed to be a single-threaded application. Any other course of action risks delaying thread shutdown and loses the information that the thread was interrupted - probably without finishing its task.

Similarly, the ThreadDeath exception should also be propagated. According to its JavaDoc:

If ThreadDeath is caught by a method, it is important that it be rethrown so that the thread actually dies.